gap/congruences/conglatt.gi

```diff
@@ -7,111 +7,81 @@
-"for a congruence lattice",
-[IsCongruenceLattice],
-function(latt)
-  return latt![2];
-end);
 
 InstallMethod(\[\],
-"for a congruence lattice",
-[IsCongruenceLattice, IsObject],
-function(latt, x)
-  return latt![1][x];
+"for a congruence poset",
+[IsCongruencePoset, IsObject],
+function(poset, x)
+  return OutNeighboursOfVertex(poset!.po, x);
 end);
 
 InstallMethod(ViewObj,
-"for a congruence lattice",
-[IsCongruenceLattice],
-function(latt)
-  ViewObj(latt![1]);
+"for a congruence poset",
+[IsCongruencePoset],
+function(poset)
+  ViewObj(OutNeighbours(poset!.po));
 end);
 
 InstallMethod(PrintObj,
-"for a congruence lattice",
-[IsCongruenceLattice],
-function(latt)
-  PrintObj(latt![1]);
+"for a congruence poset",
+[IsCongruencePoset],
+function(poset)
+  PrintObj(OutNeighbours(poset!.po));
 end);
 
 InstallMethod(Length,
-"for a congruence lattice",
-[IsCongruenceLattice],
-function(latt)
-  return Length(latt![1]);
+"for a congruence poset",
+[IsCongruencePoset],
+function(poset)
+  return DigraphNrVertices(poset!.po);
 end);
 
 InstallMethod(IsBound\[\],
-"for a congruence lattice and a positive integer",
-[IsCongruenceLattice, IsPosInt],
-function(latt, x)
-  return IsBound(latt![1][x]);
+"for a congruence poset and a positive integer",
+[IsCongruencePoset, IsPosInt],
+function(poset, x)
+  return x >= 1 and x <= Length(poset);
 end);
 
-###############################################################################
-# LatticeOfXCongruences function
-###############################################################################
-#
-# This abstract function takes a semigroup 'S', a string 'type_string' and a
-# record 'record'.
-# type_string should be in ["Left", "Right", ""] and describes the sort of
-# relations we want to find (respectively: left congruences, right congruences,
-# two-sided congruences), referred to as 'x congs' below.
-# record may contain any of the following components, which should be set to
-# 'true' to have the stated effect:
-#   * minimal - Return only minimal x-congs
-#   * 1gen - Return only x-congs with a single generating pair
-#   * restriction - Return only x-congs generated by relating these elements
-#
-###############################################################################
-
-# FIXME Monolithic functions = BAD!!! JDM
-
-SEMIGROUPS.LatticeOfXCongruences := function(S, type_string, record)
-  local report, _XSemigroupCongruence, elms, pairs, total, congs1, nrcongs,
-        children, parents, last_collected, nr, pair, badcong, newchildren,
-        newparents, newcong, i, c, p, congs, set_func, lattice, join_func,
-        length, found, ignore, start, j, k;
-
+SEMIGROUPS.PrincipalXCongruencePoset := function(S, restriction, cong_func)
   if not IsFinite(S) then
-    ErrorNoReturn("Semigroups: SEMIGROUPS.LatticeOfXCongruences: usage,\n",
+    ErrorNoReturn("Semigroups: SEMIGROUPS.PrincipalXCongruencePoset: usage,\n",
                   "first argument <S> must be a finite semigroup,");
   fi;
```

Comment:
Instead of this cryptic error message perhaps you could add an argument to this function which is the name of the function calling this one, i.e. `SemigroupCongruence` or whatever. Also I think everywhere you should be checking if `IsEnumerableSemigroupRep(S) and IsFinite(S)` since the methods will only apply to semigroups in `IsEnumerableSemigroupRep`